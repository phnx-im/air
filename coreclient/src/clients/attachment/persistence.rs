// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use sqlx::SqliteExecutor;
use uuid::Uuid;

use crate::ConversationId;

/// A record of an attachment.
///
/// Content is intentially not included in this struct.
pub(super) struct AttachmentRecord {
    pub(super) attachment_id: Uuid,
    pub(super) conversation_id: ConversationId,
    pub(super) content_type: String,
}

/// Additional information about an attachment when it is an image.
///
/// Thumbnail content is intentially not included in this struct.
pub(super) struct AttachmentImageRecord {
    pub(super) attachment_id: Uuid,
    pub(super) blurhash: String,
    pub(super) width: u32,
    pub(super) height: u32,
}

impl AttachmentRecord {
    pub(super) async fn store(
        &self,
        executor: impl SqliteExecutor<'_>,
        content: impl AsRef<[u8]>,
    ) -> anyhow::Result<()> {
        let content = content.as_ref();
        sqlx::query!(
            r#"
                INSERT INTO attachments (
                    attachment_id,
                    conversation_id,
                    content_type,
                    content
                ) VALUES (?, ?, ?, ?)
                "#,
            self.attachment_id,
            self.conversation_id,
            self.content_type,
            content,
        )
        .execute(executor)
        .await?;
        Ok(())
    }
}

impl AttachmentImageRecord {
    pub(super) async fn store(&self, executor: impl SqliteExecutor<'_>) -> anyhow::Result<()> {
        sqlx::query!(
            r#"
                INSERT INTO attachment_images (
                    attachment_id,
                    blurhash,
                    width,
                    height
                ) VALUES (?, ?, ?, ?)
                "#,
            self.attachment_id,
            self.blurhash,
            self.width,
            self.height,
        )
        .execute(executor)
        .await?;
        Ok(())
    }
}
