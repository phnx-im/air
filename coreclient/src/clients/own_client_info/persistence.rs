// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use sqlx::query;

use super::OwnClientInfo;

impl OwnClientInfo {
    pub(crate) async fn store(&self, executor: impl sqlx::SqliteExecutor<'_>) -> sqlx::Result<()> {
        let as_client_id = self.user_id.uuid();
        let domain = self.user_id.domain();
        query!(
            "INSERT INTO own_client_info (
                server_url,
                qs_user_id,
                qs_client_id,
                user_uuid,
                user_domain
            ) VALUES (?, ?, ?, ?, ?)",
            self.server_url,
            self.qs_user_id,
            self.qs_client_id,
            as_client_id,
            domain,
        )
        .execute(executor)
        .await?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use phnxtypes::identifiers::{QsClientId, QsUserId, UserId};
    use sqlx::{Row, SqlitePool};
    use uuid::Uuid;

    use super::*;

    #[sqlx::test]
    async fn store(pool: SqlitePool) -> anyhow::Result<()> {
        let own_client_info = OwnClientInfo {
            server_url: "https://localhost".to_string(),
            qs_user_id: QsUserId::random(),
            qs_client_id: QsClientId::random(&mut rand::thread_rng()),
            user_id: UserId::new(Uuid::new_v4(), "localhost".parse().unwrap()),
        };

        own_client_info.store(&pool).await?;

        let row = sqlx::query("SELECT * FROM own_client_info")
            .fetch_one(&pool)
            .await?;
        let server_url = row.try_get(0)?;
        let qs_user_id = row.try_get(1)?;
        let qs_client_id = row.try_get(2)?;
        let user_uuid = row.try_get(3)?;
        let user_domain = row.try_get(4)?;
        let loaded = OwnClientInfo {
            server_url,
            qs_user_id,
            qs_client_id,
            user_id: UserId::new(user_uuid, user_domain),
        };

        assert_eq!(loaded, own_client_info);

        Ok(())
    }
}
