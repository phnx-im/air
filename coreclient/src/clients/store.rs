// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use anyhow::bail;
use phnxtypes::messages::push_token::PushToken;

use super::{
    create_user::{
        AsRegisteredUserState, BasicUserData, PersistedUserState, PostRegistrationInitState,
        QsRegisteredUserState, UnfinalizedRegistrationState,
    },
    *,
};

/// WARNING: This enum is stored in sqlite as a blob. If any changes are made to
/// this enum, a new version in `StorableUserCreationState` must be created.
#[derive(Serialize, Deserialize)]
pub(crate) enum UserCreationState {
    BasicUserData(BasicUserData),
    InitialUserState(InitialUserState),
    PostRegistrationInitState(PostRegistrationInitState),
    UnfinalizedRegistrationState(UnfinalizedRegistrationState),
    AsRegisteredUserState(AsRegisteredUserState),
    QsRegisteredUserState(QsRegisteredUserState),
    FinalUserState(PersistedUserState),
}

impl UserCreationState {
    pub(super) fn client_id(&self) -> &AsClientId {
        match self {
            Self::BasicUserData(state) => state.client_id(),
            Self::InitialUserState(state) => state.client_id(),
            Self::PostRegistrationInitState(state) => state.client_id(),
            Self::UnfinalizedRegistrationState(state) => state.client_id(),
            Self::AsRegisteredUserState(state) => state.client_id(),
            Self::QsRegisteredUserState(state) => state.client_id(),
            Self::FinalUserState(state) => state.client_id(),
        }
    }

    pub(super) fn server_url(&self) -> &str {
        match self {
            Self::BasicUserData(state) => state.server_url(),
            Self::InitialUserState(state) => state.server_url(),
            Self::PostRegistrationInitState(state) => state.server_url(),
            Self::UnfinalizedRegistrationState(state) => state.server_url(),
            Self::AsRegisteredUserState(state) => state.server_url(),
            Self::QsRegisteredUserState(state) => state.server_url(),
            Self::FinalUserState(state) => state.server_url(),
        }
    }

    pub(super) fn new(
        client_db_connection: &Connection,
        phnx_db_connection: &Connection,
        as_client_id: AsClientId,
        server_url: impl ToString,
        password: &str,
        push_token: Option<PushToken>,
    ) -> Result<Self> {
        let client_record = ClientRecord::new(as_client_id.clone());
        client_record.store(phnx_db_connection)?;

        let basic_user_data = BasicUserData {
            as_client_id: as_client_id.clone(),
            server_url: server_url.to_string(),
            password: password.to_string(),
            push_token,
        };

        // Create user profile entry for own user.
        UserProfile::store_own_user_profile(
            client_db_connection,
            &mut StoreNotifier::noop(),
            as_client_id.user_name(),
            None,
            None,
        )?;

        let user_creation_state = UserCreationState::BasicUserData(basic_user_data);

        user_creation_state.store(client_db_connection)?;

        Ok(user_creation_state)
    }

    pub(super) async fn step(
        self,
        phnx_db_connection: SqliteConnection,
        client_db_connection: SqliteConnection,
        api_clients: &ApiClients,
    ) -> Result<Self> {
        // If we're already in the final state, there is nothing to do.
        if matches!(self, UserCreationState::FinalUserState(_)) {
            return Ok(self);
        }

        let new_state = match self {
            UserCreationState::BasicUserData(state) => Self::InitialUserState(
                state
                    .prepare_as_registration(client_db_connection.clone(), api_clients)
                    .await?,
            ),
            UserCreationState::InitialUserState(state) => {
                Self::PostRegistrationInitState(state.initiate_as_registration(api_clients).await?)
            }
            UserCreationState::PostRegistrationInitState(state) => {
                let connection = client_db_connection.lock().await;
                Self::UnfinalizedRegistrationState(state.process_server_response(&connection)?)
            }
            UserCreationState::UnfinalizedRegistrationState(state) => {
                Self::AsRegisteredUserState(state.finalize_as_registration(api_clients).await?)
            }
            UserCreationState::AsRegisteredUserState(state) => {
                Self::QsRegisteredUserState(state.register_with_qs(api_clients).await?)
            }
            UserCreationState::QsRegisteredUserState(state) => Self::FinalUserState(
                state
                    .upload_key_packages(client_db_connection.clone(), api_clients)
                    .await?,
            ),
            UserCreationState::FinalUserState(_) => self,
        };

        let client_db_connection = client_db_connection.lock().await;
        new_state.store(&client_db_connection)?;

        // If we just transitioned into the final state, we need to update the
        // client record.
        let phnx_db_connection = phnx_db_connection.lock().await;
        if let UserCreationState::FinalUserState(_) = new_state {
            let mut client_record = ClientRecord::load(&phnx_db_connection, new_state.client_id())?
                .ok_or(anyhow!("Client record not found"))?;
            client_record.finish();
            client_record.store(&phnx_db_connection)?;
        }

        Ok(new_state)
    }

    pub(super) fn final_state(self) -> Result<PersistedUserState> {
        if let UserCreationState::FinalUserState(state) = self {
            Ok(state)
        } else {
            bail!("Unexpected user creation state")
        }
    }

    /// A convenience function that performs the `step` function until the final state is reached.
    pub(super) async fn complete_user_creation(
        mut self,
        phnx_db_connection: SqliteConnection,
        client_db_connection: SqliteConnection,
        api_clients: &ApiClients,
    ) -> Result<PersistedUserState> {
        while !matches!(self, UserCreationState::FinalUserState(_)) {
            self = self
                .step(
                    phnx_db_connection.clone(),
                    client_db_connection.clone(),
                    api_clients,
                )
                .await?
        }

        self.final_state()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ClientRecordState {
    InProgress,
    Finished,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ClientRecord {
    pub as_client_id: AsClientId,
    pub client_record_state: ClientRecordState,
    pub created_at: DateTime<Utc>,
    pub is_default: bool,
}

impl ClientRecord {
    pub(super) fn new(as_client_id: AsClientId) -> Self {
        Self {
            as_client_id,
            client_record_state: ClientRecordState::InProgress,
            created_at: Utc::now(),
            is_default: false,
        }
    }

    pub(super) fn finish(&mut self) {
        self.client_record_state = ClientRecordState::Finished;
    }
}
