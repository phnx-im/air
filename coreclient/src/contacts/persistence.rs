// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use aircommon::{
    crypto::{
        ear::keys::WelcomeAttributionInfoEarKey, indexed_aead::keys::UserProfileKeyIndex,
        kdf::keys::ConnectionKey,
    },
    identifiers::{Fqdn, UserHandle, UserId},
    messages::FriendshipToken,
};
use chrono::Utc;
use sqlx::{SqliteExecutor, SqliteTransaction, query, query_as};
use tokio_stream::StreamExt;
use uuid::Uuid;

use crate::{
    Contact, ConversationId, clients::connection_offer::FriendshipPackage, store::StoreNotifier,
};

use super::HandleContact;

struct SqlContact {
    user_uuid: Uuid,
    user_domain: Fqdn,
    conversation_id: ConversationId,
    wai_ear_key: WelcomeAttributionInfoEarKey,
    friendship_token: FriendshipToken,
    connection_key: ConnectionKey,
    user_profile_key_index: UserProfileKeyIndex,
}

impl From<SqlContact> for Contact {
    fn from(
        SqlContact {
            user_uuid,
            user_domain,
            wai_ear_key,
            friendship_token,
            conversation_id,
            connection_key,
            user_profile_key_index,
        }: SqlContact,
    ) -> Self {
        Self {
            user_id: UserId::new(user_uuid, user_domain),
            wai_ear_key,
            friendship_token,
            connection_key,
            conversation_id,
            user_profile_key_index,
        }
    }
}

impl Contact {
    pub(crate) async fn load(
        executor: impl SqliteExecutor<'_>,
        user_id: &UserId,
    ) -> sqlx::Result<Option<Self>> {
        let uuid = user_id.uuid();
        let domain = user_id.domain();
        query_as!(
            SqlContact,
            r#"SELECT
                user_uuid AS "user_uuid: _",
                user_domain AS "user_domain: _",
                conversation_id AS "conversation_id: _",
                wai_ear_key AS "wai_ear_key: _",
                friendship_token AS "friendship_token: _",
                connection_key AS "connection_key: _",
                user_profile_key_index AS "user_profile_key_index: _"
            FROM contacts WHERE user_uuid = ? AND user_domain = ?"#,
            uuid,
            domain
        )
        .fetch_optional(executor)
        .await
        .map(|res| res.map(From::from))
    }

    pub(crate) async fn load_all(executor: impl SqliteExecutor<'_>) -> sqlx::Result<Vec<Self>> {
        query_as!(
            SqlContact,
            r#"SELECT
                user_uuid AS "user_uuid: _",
                user_domain AS "user_domain: _",
                conversation_id AS "conversation_id: _",
                wai_ear_key AS "wai_ear_key: _",
                friendship_token AS "friendship_token: _",
                connection_key AS "connection_key: _",
                user_profile_key_index AS "user_profile_key_index: _"
            FROM contacts"#
        )
        .fetch(executor)
        .map(|res| res.map(From::from))
        .collect()
        .await
    }

    pub(crate) async fn upsert(
        &self,
        executor: impl SqliteExecutor<'_>,
        notifier: &mut StoreNotifier,
    ) -> sqlx::Result<()> {
        let uuid = self.user_id.uuid();
        let domain = self.user_id.domain();
        query!(
            "INSERT OR REPLACE INTO contacts (
                user_uuid,
                user_domain,
                conversation_id,
                wai_ear_key,
                friendship_token,
                connection_key,
                user_profile_key_index
            ) VALUES (?, ?, ?, ?, ?, ?, ?)",
            uuid,
            domain,
            self.conversation_id,
            self.wai_ear_key,
            self.friendship_token,
            self.connection_key,
            self.user_profile_key_index,
        )
        .execute(executor)
        .await?;
        notifier
            .add(self.user_id.clone())
            .update(self.conversation_id);
        Ok(())
    }

    pub(crate) async fn update_user_profile_key_index(
        executor: impl SqliteExecutor<'_>,
        user_id: &UserId,
        key_index: &UserProfileKeyIndex,
    ) -> sqlx::Result<()> {
        let uuid = user_id.uuid();
        let domain = user_id.domain();
        query!(
            "UPDATE contacts SET user_profile_key_index = ?
            WHERE user_uuid = ? AND user_domain = ?",
            key_index,
            uuid,
            domain,
        )
        .execute(executor)
        .await?;
        Ok(())
    }
}

impl HandleContact {
    pub(crate) async fn upsert(
        &self,
        executor: impl SqliteExecutor<'_>,
        notifier: &mut StoreNotifier,
    ) -> sqlx::Result<()> {
        let created_at = Utc::now();
        query!(
            "INSERT OR REPLACE INTO user_handle_contacts (
                user_handle,
                conversation_id,
                friendship_package_ear_key,
                created_at
            ) VALUES (?, ?, ?, ?)",
            self.handle,
            self.conversation_id,
            self.friendship_package_ear_key,
            created_at,
        )
        .execute(executor)
        .await?;
        notifier.update(self.conversation_id);
        Ok(())
    }

    pub(crate) async fn load(
        executor: impl SqliteExecutor<'_>,
        handle: &UserHandle,
    ) -> sqlx::Result<Option<Self>> {
        query_as!(
            Self,
            r#"SELECT
                user_handle AS "handle: _",
                conversation_id AS "conversation_id: _",
                friendship_package_ear_key AS "friendship_package_ear_key: _"
            FROM user_handle_contacts
            WHERE user_handle = ?"#,
            handle,
        )
        .fetch_optional(executor)
        .await
    }

    pub(crate) async fn load_all(executor: impl SqliteExecutor<'_>) -> sqlx::Result<Vec<Self>> {
        query_as!(
            Self,
            r#"SELECT
                user_handle AS "handle: _",
                conversation_id AS "conversation_id: _",
                friendship_package_ear_key AS "friendship_package_ear_key: _"
            FROM user_handle_contacts"#,
        )
        .fetch_all(executor)
        .await
    }

    async fn delete(&self, executor: impl SqliteExecutor<'_>) -> sqlx::Result<()> {
        query!(
            "DELETE FROM user_handle_contacts WHERE user_handle = ?",
            self.handle
        )
        .execute(executor)
        .await?;
        Ok(())
    }

    /// Creates and persists a [`Contact`] from this [`HandleContact`] and the additional data
    pub(crate) async fn mark_as_complete(
        self,
        txn: &mut SqliteTransaction<'_>,
        notifier: &mut StoreNotifier,
        user_id: UserId,
        friendship_package: FriendshipPackage,
        user_profile_key_index: UserProfileKeyIndex,
    ) -> anyhow::Result<Contact> {
        let contact = Contact {
            user_id,
            conversation_id: self.conversation_id,
            wai_ear_key: friendship_package.wai_ear_key,
            friendship_token: friendship_package.friendship_token,
            connection_key: friendship_package.connection_key,
            user_profile_key_index,
        };

        self.delete(txn.as_mut()).await?;
        contact.upsert(txn.as_mut(), notifier).await?;

        Ok(contact)
    }
}

#[cfg(test)]
mod tests {
    use aircommon::{
        crypto::{
            ear::keys::{FriendshipPackageEarKey, WelcomeAttributionInfoEarKey},
            indexed_aead::keys::UserProfileKey,
            kdf::keys::ConnectionKey,
        },
        messages::FriendshipToken,
    };
    use sqlx::SqlitePool;

    use crate::{
        ConversationId, conversations::persistence::tests::test_conversation,
        key_stores::indexed_keys::StorableIndexedKey,
    };

    use super::*;

    fn test_contact(conversation_id: ConversationId) -> (Contact, UserProfileKey) {
        let user_id = UserId::random("localhost".parse().unwrap());
        let user_profile_key = UserProfileKey::random(&user_id).unwrap();
        let contact = Contact {
            user_id,
            wai_ear_key: WelcomeAttributionInfoEarKey::random().unwrap(),
            friendship_token: FriendshipToken::random().unwrap(),
            connection_key: ConnectionKey::random().unwrap(),
            conversation_id,
            user_profile_key_index: user_profile_key.index().clone(),
        };
        (contact, user_profile_key)
    }

    #[sqlx::test]
    async fn contact_store_load(pool: SqlitePool) -> anyhow::Result<()> {
        let mut store_notifier = StoreNotifier::noop();

        let conversation = test_conversation();
        conversation
            .store(pool.acquire().await?.as_mut(), &mut store_notifier)
            .await?;

        let (contact, user_profile_key) = test_contact(conversation.id());
        user_profile_key.store(&pool).await?;
        contact.upsert(&pool, &mut store_notifier).await?;

        let loaded = Contact::load(&pool, &contact.user_id).await?.unwrap();
        assert_eq!(loaded, contact);

        Ok(())
    }

    #[sqlx::test]
    async fn handle_contact_upsert_load(pool: SqlitePool) -> anyhow::Result<()> {
        let mut store_notifier = StoreNotifier::noop();
        let conversation = test_conversation();
        conversation
            .store(pool.acquire().await?.as_mut(), &mut store_notifier)
            .await?;

        let handle = UserHandle::new("ellie_".to_owned()).unwrap();
        let handle_contact = HandleContact {
            handle: handle.clone(),
            conversation_id: conversation.id(),
            friendship_package_ear_key: FriendshipPackageEarKey::random().unwrap(),
        };

        handle_contact.upsert(&pool, &mut store_notifier).await?;

        let loaded = HandleContact::load(&pool, &handle).await?.unwrap();
        assert_eq!(loaded, handle_contact);

        Ok(())
    }

    #[sqlx::test]
    async fn handle_contact_mark_as_complete(pool: SqlitePool) -> anyhow::Result<()> {
        let mut store_notifier = StoreNotifier::noop();
        let conversation = test_conversation();
        conversation
            .store(pool.acquire().await?.as_mut(), &mut store_notifier)
            .await?;

        let handle = UserHandle::new("ellie_".to_owned()).unwrap();
        let handle_contact = HandleContact {
            handle: handle.clone(),
            conversation_id: conversation.id(),
            friendship_package_ear_key: FriendshipPackageEarKey::random().unwrap(),
        };

        let user_id = UserId::random("localhost".parse().unwrap());
        let user_profile_key = UserProfileKey::random(&user_id)?;
        user_profile_key.store(&pool).await?;

        let friendship_package = FriendshipPackage {
            friendship_token: FriendshipToken::random().unwrap(),
            connection_key: ConnectionKey::random().unwrap(),
            wai_ear_key: WelcomeAttributionInfoEarKey::random().unwrap(),
            user_profile_base_secret: user_profile_key.base_secret().clone(),
        };

        let mut txn = pool.begin().await?;

        let contact = handle_contact
            .mark_as_complete(
                &mut txn,
                &mut store_notifier,
                user_id,
                friendship_package,
                user_profile_key.index().clone(),
            )
            .await?;

        txn.commit().await?;

        let loaded_handle_contact = HandleContact::load(&pool, &handle).await?;
        assert!(loaded_handle_contact.is_none());

        let loaded_contact = Contact::load(&pool, &contact.user_id).await?.unwrap();
        assert_eq!(loaded_contact, contact);

        Ok(())
    }

    #[sqlx::test]
    async fn handle_contact_delete(pool: SqlitePool) -> anyhow::Result<()> {
        let mut store_notifier = StoreNotifier::noop();
        let conversation = test_conversation();
        conversation
            .store(pool.acquire().await?.as_mut(), &mut store_notifier)
            .await?;

        let handle = UserHandle::new("ellie_".to_owned()).unwrap();
        let handle_contact = HandleContact {
            handle: handle.clone(),
            conversation_id: conversation.id(),
            friendship_package_ear_key: FriendshipPackageEarKey::random().unwrap(),
        };

        handle_contact.upsert(&pool, &mut store_notifier).await?;

        let mut txn = pool.begin().await?;
        handle_contact.delete(txn.as_mut()).await?;
        txn.commit().await?;

        let loaded = HandleContact::load(&pool, &handle).await?;
        assert!(loaded.is_none());

        Ok(())
    }

    #[sqlx::test]
    async fn handle_contact_upsert_idempotent(pool: SqlitePool) -> anyhow::Result<()> {
        let mut store_notifier = StoreNotifier::noop();
        let conversation = test_conversation();
        conversation
            .store(pool.acquire().await?.as_mut(), &mut store_notifier)
            .await?;

        let handle = UserHandle::new("ellie_".to_owned()).unwrap();
        let handle_contact = HandleContact {
            handle: handle.clone(),
            conversation_id: conversation.id(),
            friendship_package_ear_key: FriendshipPackageEarKey::random().unwrap(),
        };

        handle_contact.upsert(&pool, &mut store_notifier).await?;
        handle_contact.upsert(&pool, &mut store_notifier).await?; // Upsert again

        let loaded = HandleContact::load(&pool, &handle).await?.unwrap();
        assert_eq!(loaded, handle_contact);

        Ok(())
    }
}
