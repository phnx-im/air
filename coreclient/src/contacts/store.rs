// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use tls_codec::Serialize;

use crate::utils::persistance::{DataType, Persistable};

use super::*;

impl Persistable for Contact {
    type Key = UserName;
    type SecondaryKey = UserName;
    const DATA_TYPE: DataType = DataType::Contact;

    fn own_client_id_bytes(&self) -> Vec<u8> {
        self.own_client_id.tls_serialize_detached().unwrap()
    }

    fn rowid(&self) -> Option<i64> {
        self.rowid
    }

    fn key(&self) -> &Self::Key {
        &self.user_name
    }

    fn secondary_key(&self) -> &Self::SecondaryKey {
        &self.user_name
    }

    fn set_rowid(&mut self, rowid: i64) {
        self.rowid = Some(rowid);
    }
}
