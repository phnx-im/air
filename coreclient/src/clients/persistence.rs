// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use phnxtypes::{codec::PhnxCodec, identifiers::AsClientId};
use rusqlite::{params, types::FromSql, OptionalExtension, ToSql};
use serde::{Deserialize, Serialize};
use sqlx::{
    encode::IsNull, error::BoxDynError, query, query_scalar, sqlite::SqliteTypeInfo, Database,
    Decode, Encode, Sqlite, SqlitePool, Type,
};

use crate::utils::persistence::{open_phnx_db, Storable};

use super::store::{ClientRecord, ClientRecordState, UserCreationState};

// When adding a variant to this enum, the new variant must be called
// `CurrentVersion` and the current version must be renamed to `VX`, where `X`
// is the next version number. The content type of the old `CurrentVersion` must
// be renamed and otherwise preserved to ensure backwards compatibility.
#[derive(Serialize, Deserialize)]
enum StorableUserCreationState {
    CurrentVersion(UserCreationState),
}

// Only change this enum in tandem with its non-Ref variant.
#[derive(Serialize)]
enum StorableUserCreationStateRef<'a> {
    CurrentVersion(&'a UserCreationState),
}

impl Type<Sqlite> for UserCreationState {
    fn type_info() -> SqliteTypeInfo {
        <Vec<u8> as Type<Sqlite>>::type_info()
    }
}

impl<'q> Encode<'q, Sqlite> for UserCreationState {
    fn encode_by_ref(
        &self,
        buf: &mut <Sqlite as Database>::ArgumentBuffer<'q>,
    ) -> Result<IsNull, BoxDynError> {
        let state = StorableUserCreationStateRef::CurrentVersion(self);
        let bytes = PhnxCodec::to_vec(&state)?;
        <Vec<u8> as Encode<Sqlite>>::encode(bytes, buf)
    }
}

impl<'r> Decode<'r, Sqlite> for UserCreationState {
    fn decode(value: <Sqlite as Database>::ValueRef<'r>) -> Result<Self, BoxDynError> {
        let bytes = <&[u8] as Decode<'r, Sqlite>>::decode(value)?;
        let state = PhnxCodec::from_slice(bytes)?;
        match state {
            StorableUserCreationState::CurrentVersion(state) => Ok(state),
        }
    }
}

impl FromSql for UserCreationState {
    fn column_result(value: rusqlite::types::ValueRef) -> rusqlite::types::FromSqlResult<Self> {
        let state = PhnxCodec::from_slice(value.as_blob()?)?;
        match state {
            StorableUserCreationState::CurrentVersion(state) => Ok(state),
        }
    }
}

impl ToSql for UserCreationState {
    fn to_sql(&self) -> rusqlite::Result<rusqlite::types::ToSqlOutput<'_>> {
        let state = StorableUserCreationStateRef::CurrentVersion(self);
        let bytes = PhnxCodec::to_vec(&state)?;

        Ok(rusqlite::types::ToSqlOutput::from(bytes))
    }
}

impl Storable for UserCreationState {
    const CREATE_TABLE_STATEMENT: &'static str = "
        CREATE TABLE IF NOT EXISTS user_creation_state (
            client_id BLOB PRIMARY KEY,
            state BLOB NOT NULL
        );";

    fn from_row(row: &rusqlite::Row) -> anyhow::Result<Self, rusqlite::Error> {
        row.get(0)
    }
}

impl UserCreationState {
    pub(super) fn load(
        connection: &rusqlite::Connection,
        client_id: &AsClientId,
    ) -> Result<Option<Self>, rusqlite::Error> {
        connection
            .query_row(
                "SELECT state FROM user_creation_state WHERE client_id = ?1",
                [client_id],
                Self::from_row,
            )
            .optional()
    }

    pub(super) async fn load_2(
        db: &SqlitePool,
        client_id: &AsClientId,
    ) -> sqlx::Result<Option<Self>> {
        query_scalar!(
            r#"SELECT state AS "state: _" FROM user_creation_state WHERE client_id = ?"#,
            client_id
        )
        .fetch_optional(db)
        .await
    }

    pub(super) fn store(&self, connection: &rusqlite::Connection) -> Result<(), rusqlite::Error> {
        connection.execute(
            "INSERT OR REPLACE INTO user_creation_state (client_id, state) VALUES (?1, ?2)",
            params![self.client_id(), self],
        )?;
        Ok(())
    }

    pub(super) async fn store_2(&self, db: &SqlitePool) -> sqlx::Result<()> {
        let client_id = self.client_id();
        query!(
            "INSERT OR REPLACE INTO user_creation_state (client_id, state) VALUES (?, ?)",
            client_id,
            self
        )
        .execute(db)
        .await?;
        Ok(())
    }
}

// TODO: This is stored in a different db
impl Storable for ClientRecord {
    const CREATE_TABLE_STATEMENT: &'static str = "
        CREATE TABLE IF NOT EXISTS client_record (
            client_id BLOB PRIMARY KEY,
            record_state TEXT NOT NULL CHECK (record_state IN ('in_progress', 'finished'))
        );";

    fn from_row(row: &rusqlite::Row) -> anyhow::Result<Self, rusqlite::Error> {
        let record_state_str: String = row.get(1)?;
        let client_record_state = match record_state_str.as_str() {
            "in_progress" => ClientRecordState::InProgress,
            "finished" => ClientRecordState::Finished,
            _ => return Err(rusqlite::Error::InvalidQuery),
        };
        Ok(Self {
            as_client_id: row.get(0)?,
            client_record_state,
        })
    }
}

impl ClientRecord {
    pub fn load_all_from_phnx_db(phnx_db_path: &str) -> Result<Vec<Self>, rusqlite::Error> {
        let connection = open_phnx_db(phnx_db_path)?;
        let mut stmt = connection.prepare("SELECT * FROM client_record")?;
        let client_records = stmt
            .query_map([], Self::from_row)?
            .collect::<Result<Vec<_>, _>>()?;
        Ok(client_records)
    }

    pub fn load_all(connection: &rusqlite::Connection) -> Result<Vec<Self>, rusqlite::Error> {
        let mut stmt = connection.prepare("SELECT * FROM client_record")?;
        let client_records = stmt
            .query_map([], Self::from_row)?
            .collect::<Result<Vec<_>, _>>()?;
        Ok(client_records)
    }

    pub(super) fn load(
        connection: &rusqlite::Connection,
        client_id: &AsClientId,
    ) -> Result<Option<Self>, rusqlite::Error> {
        connection
            .query_row(
                "SELECT * FROM client_record WHERE client_id = ?1",
                [client_id],
                Self::from_row,
            )
            .optional()
    }

    pub(super) fn store(&self, connection: &rusqlite::Connection) -> Result<(), rusqlite::Error> {
        let record_state_str = match self.client_record_state {
            ClientRecordState::InProgress => "in_progress",
            ClientRecordState::Finished => "finished",
        };
        connection.execute(
            "INSERT OR REPLACE INTO client_record (client_id, record_state) VALUES (?1, ?2)",
            params![self.as_client_id, record_state_str],
        )?;
        Ok(())
    }
}
