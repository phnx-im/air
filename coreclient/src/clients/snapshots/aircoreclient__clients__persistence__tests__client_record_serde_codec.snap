---
source: coreclient/src/clients/persistence.rs
expression: "PersistenceCodec::to_vec(&record).unwrap()"
extension: cbor
snapshot_kind: binary
---
